Implement Mii Selector applet

Techie-Pi opened the pull request:
I've implemented the Mii Selector applet, I have yet to add documentation, but I would like to see your opinions on the API.
Specially around the creation and usage of the enums ``MiiConfigIndex`` and ``MiiType``.

FenrirWolf commented:
The CI failure is just from Xargo failing to build on older nightlies, so i'm gonna go ahead and merge this anyway.

Techie-Pi committed:
Initial mii selector implementation

Techie-Pi committed:
Fix bug with author and impl ``From`` for ``MiiConfigIndex``

Techie-Pi committed:
Add ``mii_selector`` example

Techie-Pi committed:
Cargo fmt

Techie-Pi commented:
Also, out of curiosity, why is [``MiiData`` an opaque type](https://github.com/rust3ds/ctru-rs/blob/9cf5e4de94e8d3711fd757e3464dd1880d91b92f/ctru-sys/bindgen.sh#L17)?
Nevermind, just saw the [issue](https://github.com/rust-lang/rust-bindgen/issues/1556), that's actually sad tbh

Techie-Pi commented:
I've been researching, and the representation of the MiiData given by the 3DS doesn't change at all and [is documented](https://www.3dbrew.org/wiki/Mii#Mii_format). The purpose of libctru's ``MiiData`` is to _imitate_ the representation so that the struct is generated directly from the 3DS' MiiData.

Instead of that, we could parse the raw data to a Rust-compatible struct. I'll do some testing

Techie-Pi commented:
Looks like it is possible, and quite simple tbh, maybe I could also add this to the PR.
Edit: Done... that was easier than what I had thought... I'll clean-up the code, fix a few things, and I'll incorporate the code into the PR
<img src="https://user-images.githubusercontent.com/48295028/197193792-0b71a4c3-1ba4-4df8-b00e-694282574e92.jpeg" height="300">

Techie-Pi committed:
Implement MiiData parsing

Techie-Pi committed:
Rename Mii Selector example

Techie-Pi committed:
Add MiiData example

Techie-Pi committed:
Clippy

Techie-Pi committed:
Remove comment

Techie-Pi committed:
Cargo fmt & clippy

Meziu commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,435 @@
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum OriginConsole {
+    ConsoleWii,
```
Remove the `Console` prefix from all variants.

Techie-Pi replied:
But Rust doesn't allow numerals at the start of a member (3DS)

Meziu replied:
True, hadn't noticed... Still, does "WiiUSwitch" actually exist as an option? I don't see it mentioned in `libctru`'s docs.

Techie-Pi replied:
The 3dbrew wiki says so.
[Ref](https://www.3dbrew.org/wiki/Mii#Mii_format) offset 0x3, bit 4-6

AzureMarker replied:
You can use `N3DS` instead

Techie-Pi replied:
> You can use N3DS instead

Just to clarify (_sorry it is late here_), ``N3DS`` instead of ``WiiUSwitch`` right?

Meziu replied:
No, write `N3DS` instead of `Console3DS`. At that point you can remove all the `Console` prefixes.

Techie-Pi replied:
Alright, that makes sense, idk why I thought that it was meant to replace ``WiiUSwitch``.
About replacing ``Console3DS`` with ``N3DS``, wouldn't people confuse ``N3DS`` with New 3DS?

AzureMarker replied:
We can use `New3DS` or `NewN3DS` for that. `N3DS` could mean "Nintendo 3DS"

AzureMarker replied:
Or use `O3DS` for "Original 3DS", but I don't think that's used often.

I named one of my repos [`n3ds-controller`](https://github.com/AzureMarker/n3ds-controller/) because of this issue.

Techie-Pi replied:
I think I'll use N3DS and clarify this in the documentation, seems like the best option to me

Meziu commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,435 @@
+    pub author_name: String,
+}
+
+impl From<[u8; 92]> for MiiData {
```
Use the `ctru-sys` type rather than `[u8;92]`. Do this for all the cases in which it appears.

Techie-Pi replied:
The problem is that [there is no ``ctru-sys`` type for ``[u8; 92]``](https://github.com/rust3ds/ctru-rs/blob/9cf5e4de94e8d3711fd757e3464dd1880d91b92f/ctru-sys/src/bindings.rs#L17597), but I could make a type alias

Meziu replied:
You should just use `ctru_sys::MiiData` for the conversion and take the inner buffer *here* rather than in `MiiSelectorResult`.

Meziu commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum RegionLock {
+    NoLock,
```
`None` would be less repetitive.

Techie-Pi committed:
Apply suggestions

Meziu commented on ctru-rs/examples/mii-selector.rs:
```diff
@@ -0,0 +1,45 @@
+
+    let mut mii_selector = MiiSelector::init();
+    mii_selector.set_initial_index(3);
+    mii_selector.blacklist_user_mii(0.into());
```
Why blacklist the user's mii?

Techie-Pi replied:
That's just an example, [the 3ds-examples' example also does this](https://github.com/devkitPro/3ds-examples/blob/cde9dff910f1115de79752cf55c609a9616cdc94/mii_selector/source/main.c#L54)

Meziu commented on ctru-rs/examples/mii-selector.rs:
```diff
@@ -0,0 +1,45 @@
+    println!("\x1b[4;0HName: {:?}", result.name());
+    println!("\x1b[5;0HAuthor: {:?}", result.author());
+    println!(
+        "\x1b[5;0HDoes it have moles?: {:?}",
```
This message and the "author" message are both on row 5. In this order the "Moles" message wipes the "Author" one.

Techie-Pi committed:
Fix Mii Selector example

Techie-Pi committed:
Fix ``RegionLock::None`` refactor

Meziu commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,193 @@
+}
+
+pub struct MiiSelectorReturn {
+    raw_return: Box<ctru_sys::MiiSelectorReturn>,
```
We shouldn't hold this pointer. It has no use after being consumed by our wrapper.

Meziu commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,193 @@
+    }
+}
+
+impl MiiSelectorReturn {
```
These functions are duplicate features already present (name and author can be obtained by the `MiiData`).

Furthermore, the checksum (while useful in a C environment) is a case we just shouldn't expect the user to check. The case of a faulty checksum means that our `MiiSelectorReturn` wrapped value is invalid, which is one of the safety measures we should prevent *by default*.

The checksum should be checked before returning the mii data to the user, and if the check fails, it should return an error. This check should be handled in the `MiiSelector::launch` method.

Meziu commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,193 @@
+    pub is_mii_selected: bool,
+    pub mii_type: MiiType,
+    pub checksum: u16,
+}
```
All of the values in this struct should be `read-only`, since this is a *result* type for our operations. Rather than being `pub`, they should each have a corresponding `get` method.

Also, as stated in another comment, the presence of `checksum` here is to be discussed (as its validity would be already checked), thought I'd like to hear reasons why it may actually be useful, since I've never played around with this part of `libctru`.

AzureMarker replied:
I think it's fine to return the whole struct and let the user mess with it. It's not the library's problem after it gets returned to the user. They can muck around with it all they want, we don't need it back :).

Techie-Pi replied:
I agree with AzureMaker, and the ``checksum`` thing has been solved

Meziu commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+
+#[derive(Clone, Debug)]
+pub struct MiiData {
+    pub mii_options: MiiDataOptions,
```
I think you should remove all the `mii_` prefixes in the struct's members. Most of the time they will be called directly from the `mii_data.[member]` signature.

AzureMarker commented on ctru-rs/examples/mii-selector.rs:
```diff
@@ -0,0 +1,45 @@
+
+    let result = mii_selector.launch();
+
+    println!("\x1b[0;0HIs Mii selected?: {:?}", result.is_mii_selected);
```
I don't think the line positioning code is needed. The console automatically handles newlines.

Edit: Running this on my 3DS, the second line overwrites the first line until you remove the line positioning code.

AzureMarker commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,193 @@
+    pub fn set_title(&mut self, text: &str) {
+        // This can only fail if the text contains NUL bytes in the string... which seems
+        // unlikely and is documented
+        let c_text = CString::new(text).expect("Cstring::new failed");
```
nit: based on the comment the `expect` error message could be more accurate/useful.

Techie-Pi replied:
Is it better now?

AzureMarker commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,193 @@
+                    ctru_sys::MIISELECTOR_GUESTMII_SLOTS,
+                )
+            },
+        }
```
nit: Could reduce duplication (this applies to multiple methods):
```suggestion
        let index = match mii_index {
            MiiConfigIndex::Index(i) => i,
            MiiConfigIndex::All => ctru_sys::MIISELECTOR_GUESTMII_SLOTS,
        };
        unsafe { ctru_sys::miiSelectorWhitelistGuestMii(self.config.as_mut(), index) };
```

Even better, move the `match` conversion to a method on `MiiConfigIndex`.

Techie-Pi replied:
> Even better, move the match conversion to a method on MiiConfigIndex.

The problem with that is that, depending on the context, ``MiiConfigIndex::All`` returns ``MIISELECTOR_GUESTMII_SLOTS`` or ``MIISELECTOR_USERMII_SLOTS``

AzureMarker replied:
In that case just doing the refactor in each method would be good.

AzureMarker commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,193 @@
+}
+
+pub struct MiiSelectorReturn {
+    raw_return: Box<ctru_sys::MiiSelectorReturn>,
```
Why include this in the struct if it's private? This struct will only be used by users, who can't access this field.

Techie-Pi replied:
Oh yeah, that's something that I left from testing, I'll remove it
Edit: Actually, it was used. But I've removed it

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+    String::from_utf16(raw_utf16_composed.as_slice())
+        .unwrap()
+        .replace('\0', "")
+}
```
nit: This could be simpler:
```suggestion
fn utf16_byte_pairs_to_string(data: &[u8]) -> String {
    let raw_utf16_composed = data
        .chunks_exact(2)
        .map(|chunk| u16::from_le_bytes([chunk[0], chunk[1]]))
        .collect::<Vec<u16>>();

    String::from_utf16_lossy(raw_utf16_composed.as_slice())
}
```

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+        .map(|v| vec_bit(data[*v]))
+        .collect::<Vec<[bool; 8]>>()
+        .concat()
+}
```
nit: This could be simpler:
```suggestion
fn get_and_concat_vec_bit(data: &[u8], get_values: &[usize]) -> Vec<bool> {
    get_values
        .iter()
        .flat_map(|v| vec_bit(data[*v]))
        .collect()
}
```

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+            get_and_concat_vec_bit(raw_mii_data.as_slice(), &[0x18, 0x19])
+                .try_into()
+                .unwrap();
+        let raw_utf16_name = &raw_mii_data.as_slice()[0x1A..0x2D];
```
This can just be `let raw_utf16_name = &raw_mii_data[0x1A..0x2D];` (`as_slice` isn't needed)

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+            raw_mii_data[0x15],
+        ];
+        let raw_details: [bool; 16] =
+            get_and_concat_vec_bit(raw_mii_data.as_slice(), &[0x18, 0x19])
```
nit: it's more idiomatic to do `&raw_mii_data` instead of `raw_mii_data.as_slice()`

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+
+        let options = MiiDataOptions {
+            is_copying_allowed: *raw_options.first().unwrap(),
+            is_profanity_flag_enabled: *raw_options.get(1).unwrap(),
```
```suggestion
            is_copying_allowed: raw_options[0],
            is_profanity_flag_enabled: raw_options[1],
```

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+                    RegionLock::USA
+                } else {
+                    RegionLock::Europe
+                }
```
Do this refactoring for the other similar instances below as well.
```suggestion
                match (raw_options[3], raw_options[2]) {
                    (false, false) => RegionLock::None,
                    (false, true) => RegionLock::Japan,
                    (true, false) => RegionLock::USA,
                    (true, true) => RegionLock::Europe,
                }
```

Techie-Pi replied:
Ohhh that's actually great! I hadn't thought of doing that

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+                    MiiSex::Female
+                } else {
+                    MiiSex::Male
+                }
```
```suggestion
                if raw_details[0] {
                    MiiSex::Female
                } else {
                    MiiSex::Male
                }
```

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+}
+
+/// The reverse allows to write things on a more _humane_ way, but return a LE u8
+fn partial_vec_to_u8_with_reverse(data: &[bool]) -> u8 {
```
This function's functionality is still unclear to me, maybe a better name and more descriptive docs would help?

Techie-Pi replied:
Alright, that function was _actually_ misused. The purpose of the function was writing BE bits, filling them until the length was 8 bits and making them LE.

This function was used this way before, but now the data is directly from the 3DS and this isn't needed + it caused some _big big_ bugs. I've fixed it and documented it

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+        result ^= bit;
+    });
+    result
+}
```
nit: This could be simplified:
```suggestion
fn vec_bit_to_u8(data: [bool; 8]) -> u8 {
    data.into_iter().fold(0, |result, bit| {
        (result << 1) ^ u8::from(bit)
    })
}
```

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,479 @@
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum RegionLock {
```
It makes sense to derive `Copy` for most of these structs/enums.

Techie-Pi committed:
Apply some suggestions

Techie-Pi committed:
Improve ``mii-selector`` example and a few fixes

Techie-Pi committed:
Clippy & cargo fmt

Techie-Pi committed:
Fix issue related to endianness and improve wording

Techie-Pi committed:
Add docs

Techie-Pi commented:
(For some reason I can't request multiple reviews at the same time /shrug)

Techie-Pi committed:
Document values sometimes not being straight forward

Techie-Pi committed:
Cargo fmt

Techie-Pi committed:
Document ``set_title`` input and NUL bytes

Techie-Pi committed:
Fix rustdoc link

Techie-Pi committed:
Move some styles inside the their details struct for consistency

AzureMarker commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,190 @@
+    pub mii_data: MiiData,
+    pub is_mii_selected: bool,
+    pub mii_type: MiiType,
+    pub checksum: u16,
```
Is the checksum relevant now that the original data structure isn't returned?

Techie-Pi replied:
Oh yeah, I forgot to remove that

AzureMarker commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,190 @@
+    }
+}
+
+impl From<Box<ctru_sys::MiiSelectorReturn>> for MiiSelectorReturn {
```
nit: this should be `From<ctru_sys::MiiSelectorReturn>`. The `Box` part isn't needed/utilized.

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,461 @@
+        let face_details = FaceDetails {
+            style: FaceStyle {
+                is_sharing_enabled: !raw_face_style[1],
+                shape: partial_u8_bits_to_u8(&raw_face_style[1..4]),
```
Are you sure `is_sharing_enabled` uses `raw_face_style[1]`? `shape` also uses that value.

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,461 @@
+            },
+            birthday_month: partial_u8_bits_to_u8(&raw_details[1..4]),
+            birthday_day: partial_u8_bits_to_u8(&raw_details[5..9]),
+            shirt_color: partial_u8_bits_to_u8(&raw_details[10..13]),
```
I see some missed bits here. `1..4` is `1, 2, 3`, and `5..9` is `5, 6, 7, 8` (4 is missing).
This shows up a lot, is it expected?

Techie-Pi replied:
Yeah this is expected, some bits are completely unused/padding according to 3dbrew

Techie-Pi replied:
Hmm now that I see that maybe you're right here, I had misread what you said :p

Techie-Pi replied:
Turns out you were completely right! The values I were using for testing just didn't trigger any bug

Meziu commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,190 @@
+
+/// Return value from a MiiSelector's launch
+#[derive(Clone, Debug)]
+pub struct MiiSelectorReturn {
```
`MiiSelectorReturn` should be `non_exhaustive` to prevent the user from creating it without using the service.

Techie-Pi committed:
Address comments

- Make ``MiiSelectorReturn`` ``non_exhaustive``
- Remove ``checksum`` from ``MiiSelectorReturn``
- Change ``From`` impl for ``MiiSelectorReturn``
- Rename ``OriginConsole`` members
- Fix ranges

Techie-Pi committed:
Apply cargo check suggestions

Apply nightly lints
I updated my toolchain to 1.66-nightly to fix all the files

Techie-Pi commented:
I've squashed some commits (mostly docs and cargo fmt/clippy)

Meziu reviewed (approved):
Looks good to me!

AzureMarker commented on ctru-rs/src/applets/mii_selector.rs:
```diff
@@ -0,0 +1,188 @@
+    fn from(ret: ctru_sys::MiiSelectorReturn) -> Self {
+        let raw_mii_data = ret.mii;
+        let no_mii_selected = ret.no_mii_selected;
+        let guest_mii_index_clone = ret.guest_mii_index;
```
nit: This variable isn't needed, you can directly reference `ret.guest_mii_index`

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,462 @@
+            style: partial_u8_bits_to_u8(&raw_eyebrow_details[0..=4]),
+            color: partial_u8_bits_to_u8(&raw_eyebrow_details[5..=7]),
+            scale: partial_u8_bits_to_u8(&raw_eyebrow_details[8..=11]),
+            y_scale: partial_u8_bits_to_u8(&raw_eyebrow_details[12..=14]),
+            rotation: partial_u8_bits_to_u8(&raw_eyebrow_details[16..=19]),
+            x_spacing: partial_u8_bits_to_u8(&raw_eyebrow_details[21..=24]),
```
Are there supposed to be gaps here?

Techie-Pi replied:
Yes, here it is

Techie-Pi replied:
![IMG_20221102_072428.jpg](https://user-images.githubusercontent.com/48295028/199413744-13a8a30d-80d2-42f6-a20f-518bd5eb1750.jpg)

AzureMarker commented on ctru-rs/src/mii.rs:
```diff
@@ -0,0 +1,462 @@
+    let leading_zeroes_to_add = 8 - data.len();
+    let leading_zeroes = vec![false; leading_zeroes_to_add];
+
+    vec_bit_to_u8([data, &leading_zeroes].concat().try_into().unwrap())
```
If these are leading zeros, why aren't they at the start of the list?

Techie-Pi replied:
Because of the way the bits are ordered by the 3ds, (this time I tested thoroughly to make sure the values returned are correct, so this should always work :p, but I'll double check later)

Techie-Pi replied:
Yeah. Tested again and it works properly

Techie-Pi committed:
Address comments

ian-h-chamberlain reviewed (approved):
Sorry I meant to review this a while ago, but finally got a chance to take a look and it seems pretty cool! 

Some of `ctru::MiiData` might be worth documenting further, if we're going to maintain a totally separate struct vs the `ctru_sys::MiiData`, but I think this is good to start. There's lots of documentation stuff that needs to be done across the board so I don't think it should keep this from merging as-is.

Techie-Pi commented:
After current concerns about this PR are resolved, I plan on adding (at least) some basic documentation (and maybe references to the 3dbrew wiki) before merging, though I think the best thing would be to merge this PR with basic documentation and add more detailed documentation in a follow-up PR (I'm not a native English speaker, so I may need help in the follow-up PR, which may take some time to make because of that :p)

Techie-Pi committed:
Rename a few variables and document skipped bits

Techie-Pi commented:
Is there anything that should be documented in this PR instead of the follow-up? Or is this ready to merge?

Meziu commented:
> Is there anything that should be documented in this PR instead of the follow-up? Or is this ready to merge?

I believe it’s fine for now. PR anything you want to add.
